Math/NumberTheory/ArithmeticFunctions/Inverse.hs

```diff
@@ -0,0 +1,363 @@
+atomicSeries point (ArithmeticFunction f g) (PrimePowers p ks) =
+  DS.fromDistinctAscList (map (\k -> (g (f p k), point (unPrime p ^ k))) ks)
+
+-- | See section 5.1 of the paper.
+invTotient
+  :: forall a. (Num a, UniqueFactorisation a, Eq a)
+  => [(Prime a, Word)]
+  -- ^ Factorisation of a value of the totient function
+  -> [PrimePowers a]
+  -- ^ Possible prime factors of an argument of the totient function
+invTotient fs = map (\p -> PrimePowers p (doPrime p)) ps
+  where
+    divs :: [a]
+    divs = runFunctionOnFactors divisorsListA fs
+
+    ps :: [Prime a]
+    ps = mapMaybe (isPrime . (+ 1)) divs
+
+    doPrime :: Prime a -> [Word]
+    doPrime p = case lookup p fs of
+      Nothing -> [1]
+      Just k  -> [1 .. k+1]
+
+-- | See section 5.2 of the paper.
+invSigma
+  :: forall a. (Euclidean a, Integral a, UniqueFactorisation a)
+  => [(Prime a, Word)]
+  -- ^ Factorisation of a value of the sum-of-divisors function
+  -> [PrimePowers a]
+  -- ^ Possible prime factors of an argument of the sum-of-divisors function
+invSigma fs
+  = map (\(x, ys) -> PrimePowers x (S.toList ys))
+  $ M.assocs
+  $ M.unionWith (<>) pksSmall pksLarge
+  where
+    numDivs :: a
+    numDivs = runFunctionOnFactors tauA fs
+
+    divs :: [a]
+    divs = runFunctionOnFactors divisorsListA fs
+
+    n :: a
+    n = product $ map (\(p, k) -> unPrime p ^ k) fs
+
+    -- There are two possible strategies to find possible prime factors
+    -- of an argument of the sum-of-divisors function.
+    -- 1. Take each prime p and each power e such that p^e <= n,
+    -- compute sigma(p^e) and check whether it is a divisor of n.
+    -- (corresponds to 'pksSmall' below)
+    -- 2. Take each divisor d of n and each power e such that e <= log_2 d,
+    -- compute p = floor(e-th root of (d - 1)) and check whether sigma(p^e) = d
+    -- and p is actually prime (correposnds to 'pksLarge' below).
+    --
+    -- Asymptotically the second strategy is beneficial, but computing
+    -- exact e-th roots of huge integers (especially when they exceed MAX_DOUBLE)
+    -- is very costly. That is why we employ the first strategy for primes
+    -- below limit 'lim' and the second one for larger ones. This allows us
+    -- to loop over e <= log_lim d which is much smaller than log_2 d.
+    --
+    -- The value of 'lim' below was chosen heuristically;
+    -- it may be tuned in future in accordance with new experimental data.
+    lim :: a
+    lim = numDivs `max` 2
+
+    pksSmall :: Map (Prime a) (S.Set Word)
+    pksSmall
+      = M.fromDistinctAscList
+      $ filter (not . null . snd)
+      $ map (\p -> (p, doPrime p))
+      $ takeWhile ((< lim) . unPrime)
+      $ primes
+
+    doPrime :: Prime a -> S.Set Word
+    doPrime (unPrime -> p) = S.fromDistinctAscList
+      [ e
+      | e <- [1 .. intToWord (integerLogBase (toInteger p) (toInteger n))]
+      , n `rem` ((p ^ (e + 1) - 1) `quot` (p - 1)) == 0
+      ]
+
+    pksLarge :: M.Map (Prime a) (S.Set Word)
```

Comment:
Very minor point: `M.Map` here can be replaced with `Map`.

Reply:
More minor still: isn't it more consistent to import the `Set` type unqualified to be same as `Map`?

Reply:
Done, thanks.